File: go/hot_100/4_median_of_two_sorted_arrays.go
package main

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m := len(nums1)
	n := len(nums2)
	if m > n {
		return findMedianSortedArrays(nums2, nums1)
	}
	left, right := 0, m
	for left <= right {
		i := (left + right) / 2
		j := (m+n+1)/2 - i
		if i != 0 && nums1[i-1] > nums2[j] {
			right = i - 1
		} else if i != m && nums2[j-1] > nums1[i] {
			left = i + 1
		} else {
			var max_left, min_right int
			if i == 0 {
				max_left = nums2[j-1]
			} else if j == 0 {
				max_left = nums1[i-1]
			} else {
				max_left = int(math.Max(float64(nums1[i-1]), float64(nums2[j-1])))
			}
			if (m+n)%2 == 1 {
				return float64(max_left)
			}
			if i == m {
				min_right = nums2[j]
			} else if j == n {
				min_right = nums1[i]
			} else {
				min_right = int(math.Min(float64(nums1[i]), float64(nums2[j])))
			}
			return float64(max_left+min_right) / 2
		}
	}
	return 0.0
}

// func main() {
// 	findMedianSortedArrays([]int{1, 3}, []int{2, 7})
// }

File: python/hot_100/4_median-of-two-sorted-arrays.py
from typing import List
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        left,right=0,len(nums1)
        m,n=len(nums1),len(nums2)
        if m>n:
            return self.findMedianSortedArrays(nums2,nums1)
        # if not m:
        #     midean1=nums2[(n-1)//2]
        #     midean2=nums2[n//2]
        while left<=right:
            i=(left+right)//2        
            j=(m+n+1)//2-i
            nums1im=nums1[i-1] if i!=0 else -float('inf')
            nums1i=nums1[i] if i!=m else float('inf')
            nums2jm=nums2[j-1] if j!=0 else -float('inf')
            nums2j=nums2[j] if j!=n else float('inf')

            if nums1im<=nums2j:
                midean1=max(nums1im,nums2jm)
                midean2=min(nums2j,nums1i)
                left=i+1
            else:
                right=i-1

        return midean1 if (m+n)%2 else (midean1+midean2)/2
        
s=Solution()
print(s.findMedianSortedArrays([1,3],[2,7]))
            